rfc/20240524-OpenTofu-RFC-Process.md

```diff
@@ -0,0 +1,84 @@
+The following will be split between [CONTRIBUTING.md](../CONTRIBUTING.md) and [rfc/README.md](./README.md):
+
+The process starts with an issue (`enhancement` or `bug`) is submitted to the repository. The `needs-rfc` label is added during the Core Team Triage process and indicates that an in-depth discussion and consensus is needed. This facilitates conversations and discussions around complex issues that are hard to have in the GitHub Issue format.
+
+If interested, a Community Member will take the following actions to submit an RFC:
+1. Copy the [./rfc/yyyymmdd-template.md](link) to `./rfc/${isodate}-${rfc title}.md` on a branch in their fork of the OpenTofu Repository
+2. Edit the newly created Markdown file and fill in the template fields
+3. Submit a Pull Request in the OpenTofu Repository, linked to the open issue(s)
+   - A Draft Pull Request is recommended if early feedback or help is needed to fully fill out the template
+4. The Community Members discuss the RFC in detail until all open questions are resolved
+5. The majority of the Core Team Approves the RFC Pull Request
+   - If a consensus is not reached, the Pull Request is closed.
+   - The Core Team may ask for a new RFC or may close the original issue entirely depending on the discussion that has happened.
+6. The RFC is Merged, and the Core Team creates issues in the relevant repositories to track the work required to implement the RFC.
+
+### Technical Approach
+
+No automation is proposed as part of this process, however we are using the RFC Template proposed in this PR to "dog-food" the process.
+
+No additional labels are required in the repository.
+
+Updates to the CONTRIBUTING.md are required as described above and the existing RFC template will need to be removed.
+
+### Open Questions
+
+* Should we use ISODATE or assigned SERIAL in the RFC file name
+  - ISODATE could either be proposed date or accepted date
+  - SERIAL is clearly linear, but requires additional processing once the PR is Accepted.
+  - For now we are using proposed date, but automation could easily be added to amend the RFC once it has been merged.
+* Should we backport some or all of the current RFCs accepted into OpenTofu to additionally validate this process?
+
+### Future Considerations
+
+#### Moving to a separate repository
+
+For now, we are proposing to keep all of the RFC files in the main OpenTofu Repository. We believe that this helps with discoverability and co-locates it with the majority of the ongoing development effort. This does have some downsides however, primarily in waiting on the required GitHub actions (testing) that are run for all pull requests. If we ever decide to change this location to either RFCs in a single separate repository or per-repository, the specific history can easily be pulled over using standard git commands.
+
+As mentioned in [Open Questions](#Open-Questions), we may build automation using GitHub Actions to post-process RFCs once they are accepted and merged.
```

Comment:
FWIW, I really prefer the idea of having a seperate repo so that we can also have RFCs for the registry and other repos.

Reply:
My preference would be to have all of the RFCs in the main repo, including the ones for the registry.  Frequently RFCs in other repos will still require updates or at least significant context in the OpenTofu repo.

Reply:
Having discussed with the team, we are going to keep these in the OpenTofu repo and keep them under the MPL.

We may move them to their own repo at some point, but will not plan on any license changes.

Additionally, as code examples may be included in RFCs, it provides a clear MPL history if they are adapted and pulled into the main code base.